agent/fw_cakephp.c

```diff
@@ -281,35 +268,40 @@ NR_PHP_WRAPPER(nr_cakephp_problem_1) {
 NR_PHP_WRAPPER_END
 
 /*
- * CakePHP 2.0+
+ * CakePHP 4.0+
  *
- * If the action or controller is not found during the dispatch process, the
- * appropriate Exception will be created and thrown. We wrap the CakeException
- * constructor instead of the Exception handler, since CakePHP allows for the
- * handler to be completely replaced.
- *
- *  CakePHP 2.x is end-of-life and in maintenance mode (critical bugfixes only).
- * As such, functionality added in PHP 7.1+ is not well supported.
- *
- * txn naming scheme:
- * In this case, `nr_txn_set_path` is called before `NR_PHP_WRAPPER_CALL` with
- * `NR_NOT_OK_TO_OVERWRITE` and as this corresponds to calling the wrapped
- * function in func_begin it needs to be explicitly set as a before_callback to
- * ensure OAPI compatibility. This entails that the first wrapped call gets to
- * name the txn.
+ * Report errors and exceptions caught by CakePHP's error handler.
  */
-NR_PHP_WRAPPER(nr_cakephp_problem_2) {
-  const char* name = "Exception";
+NR_PHP_WRAPPER(nr_cakephp_error_handler_wrapper) {
+  zval* exception = NULL;
+  char* request_uri
+      = nr_strdup(nr_php_get_server_global("REQUEST_URI" TSRMLS_CC));
 
-  (void)wraprec;
   NR_UNUSED_SPECIALFN;
+  (void)wraprec;
 
   NR_PHP_WRAPPER_REQUIRE_FRAMEWORK(NR_FW_CAKEPHP);
 
-  nr_txn_set_path("CakePHP", NRPRG(txn), name, NR_PATH_TYPE_ACTION,
+  exception = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  if (!nr_php_is_zval_valid_object(exception)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: exception is NULL or not an object",
+                     __func__);
+    goto end;
+  }
+
+  if (NR_SUCCESS
+      != nr_php_error_record_exception(
+          NRPRG(txn), exception, nr_php_error_get_priority(E_ERROR), true,
+          "Uncaught exception ", &NRPRG(exception_filters) TSRMLS_CC)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: unable to record exception", __func__);
+  }
+
+  nr_txn_set_path("CakePHP", NRPRG(txn), request_uri, NR_PATH_TYPE_ACTION,
```

Comment:
Should fourth argument be `NR_PATH_TYPE_ACTION` or `NR_PATH_TYPE_URI`? Judging by third argument, it should be `NR_PATH_TYPE_URI`.

Reply:
I believe the fourth argument should stay as `NR_PATH_TYPE_ACTION` because if it is changed to `NR_PATH_TYPE_URI`, then the `WebTransaction` metric will not include the controller name or the action name.

Reply:
But isn't the transaction named after request uri not after `Controller/method`?

Reply:
After discussion, this has been changed to `NR_PATH_TYPE_URI`. I also have added a `if` check to ensure the request uri is not `NULL`: [fa779f5](https://github.com/newrelic/newrelic-php-agent/pull/983/commits/fa779f5a01cf2ace614bfced4e68afd8dbccb1ec)